### ISSUE TITLE
Support more version notations in plugins DSL

### ISSUE THREAD

**Issue by jjohannes** [gradle/gradle#14704]
Currently, only fixed version numbers (required versions) are allowed. However, full dependency management is used now to resolve the plugins so we can support:

1. ~Anything that has a string notation, including dynamic versions~ fixed by #16046
2. Define no version at all (this is the recommendation in any case, but it is not supported in all cases yet)

For, 1. it seems to be enough to just remove the restriction here: https://github.com/gradle/gradle/pull/14703/files#diff-ea694e00176f7621b9d611f37e9a0880L69-L72

For 2. We need to allow empty versions and then we need some other way to determine that the marker exists. Then, if the version is available transitively, we would still succeed. Test case: https://github.com/gradle/gradle/pull/14703/files#diff-55c709d9faf4ed2a60c51dc13118d09cR488
With the current approach, we could use `+` as version when resolving the marker (as in the spike #14703) but that's inefficient because we'll do more requests to the repository. A better solution would probably be not to resolve a marker, but just check if any version of the marker exists in the repository.

**Comment by staale** [comment]
This is something that directly affects us. We have a large number of projects that uses our own set of plugins to configure common traits across all projects. We are currently doing this by including plugins via buildscript classpath, and use range notation (1.0.+). This allows us to improve our plugins, distribute to all our projects, and have them all upgrade on next build.

We take care to never break anything in these patch upgrades, but we may add extra features, monitoring, and fix bugs that will benefit all our projects by simply just building them again.

Since we can't use version ranges for the plugins block, we are currently stuck, and unable to update to use this feature. And looking at the code, it should not be difficult to fix this by just removing a restriction check.

**Comment by jjohannes** [comment]
First point "Anything that has a string notation, including dynamic versions" has been addressed by #16046,

"Define no version at all" is still an open question.

**Comment by tellisnz-shift** [comment]
should 

```
configurations.all {
    resolutionStrategy.cacheDynamicVersionsFor(0, "seconds")
}
```

control the caching for this or is there another way to set the dynamic version cache for plugins?

**Comment by jjohannes** [comment]
Plugin resolution is done on the `classpath` configuration in the _buidlscript_ scope. This should work I think:

```
buildscript { 
  configurations.classpath.cacheDynamicVersionsFor(0, "seconds")
}
```

**Comment by jbartok** [comment]
Fix for the first point was done in Gradle 7.0.
We'll likely not do point two ("Define no version at all").
